design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+topic *does not exist* and these values are specified, this command will create a topic with a the
+set number of partitions and replicas before executing the command. If the topic exists and has
+a different number of partitions/replicas, this command will fail.
+
+This functionality will be flagged under a configuration `ksql.ddl.source.topic.create.allow`,
+which will default to `true` (new behavior is the default).
+
+## Design
+
+### API Design
+
+The above APIs make various trade-offs, outlined below:
+
+* The proposal to overload `INSERT INTO` is to remain as compatible with SQL standard as possible.
+The parser should be lookahead and determine whether or not `VALUES` is present in order to leverage
+that data producing APIs. An alternative would be to use the word `PRODUCE`, which also makes it
+clear that it uses low-level Kafka producers.
+* The proposal to use `DELETE` encompasses various trade-offs:
+  * We need a different mechanism to produce tombstones than `INSERT INTO` because there is no good
+  way with the standard SQL to differentiate between an entirely `NULL` value (i.e. tombstone) and 
+  `NULL` values for every column.
+  * Introducing a new syntax (e.g. `INSERT TOMBSTONE`) will require users to understand another
+  language construct and make KSQL depart further from SQL standard
+  * This is a step closer to SQL standard, and we can always support more `DELETE` conditions in
+  the future
+* Allowing `CREATE TABLE/STREAM` to specify a kafka topic that does not yet exist makes it easy to
+use the new APIs without any other system. There are three alternatives: 
+  * Do not require `PARTITIONS`/`REPLICAS` to create the kafka topic. This is a slight change that
+  will make it easier to experiment, but may open the door to accidentally creating kafka topics.
+  * Introduce another `CREATE TOPIC` API which will create empty topics. I believe this will cause
+  confusion, as it departs farther from SQL standard and puts `TOPIC` on the same level as the
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. This
+operation will not be available in headless mode for the first version of the feature.
```

Comment:
So we will basically fail and exit in headless mode, or ignore the statements? 

I can see why such a feature may not make the most sense in headless mode. But are the strong reasons to bifurcate the two modes like this? I mean, if people use this statement in headless mode, and deploy the sql file on multiple servers, and get duplicates, that's probably ok. Are there any other ill effects?

Reply:
On the contrary - I do imagine some useful scenarios for this in headless mode! As of SQL-92, there is syntax for `CREATE ASSERTION` (see https://community.oracle.com/ideas/13028). With that type of behavior (combined with this) you could write a fully standalone KSQL system test 😍.

The only reason I'm drawing the distinction is that it's more work add the functionality (as opposed to less work to restrict it) in headless mode.

Reply:
I'd keep headless supporting this. I don't see any need to exclude it.